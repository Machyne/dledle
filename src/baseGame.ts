// Give these string values so they can be round trip through being used as object keys.
// These are never encoded as-is so strings are fine.
export enum GameScore {
  Win = "Win",
  NearWin = "NearWin",
  Loss = "Loss",
}

export type GameResult = {
  score: GameScore;
  serializedResult: string;
};

export abstract class BaseGame {
  private _resultRegex: RegExp | null = null;

  /**
   * The name of the game.
   */
  abstract get name(): string;

  /**
   * The name of the game in lowercase with only alphanumeric characters.
   */
  get fileName(): string {
    return this.name.toLowerCase().replace(/[^a-z0-9]/g, "");
  }

  /**
   * A regex that matches the result of the game.
   */
  get resultRegex(): RegExp {
    if (this._resultRegex === null) {
      this._resultRegex = this._buildResultRegex();
    }
    return this._resultRegex;
  }

  /**
   * Builds a regex that matches the result of the game. The return value will be cached.
   */
  abstract _buildResultRegex(): RegExp;

  /**
   * A link to the game.
   */
  abstract get link(): string;

  /**
   * Parses the user-supplied game result and returns a serialized version of
   * the game data in addition to the "score".
   *
   * @param gameResult the result of the game run through the resultRegex
   */
  abstract serializeResult(gameResult: RegExpMatchArray): GameResult;

  /**
   * The share text of the result of the game given the serialized game data.
   *
   * @param serializedResult the serialized game data
   */
  abstract deserialize(serializedResult: string): string;
}
