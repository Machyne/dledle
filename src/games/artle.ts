import { BaseGame } from "../baseGame";
import { emojiSquares } from "../util/emoji";
import { simpleGuessCountGame } from "../util/simpleGuessCountGame";

const { regex, serializeResult, deserialize } = simpleGuessCountGame({
  prefixEmoji: "ðŸŽ¨",
  maxGuesses: 4,
  gameName: "Artle",
  greySquare: emojiSquares.white,
});

export class Artle extends BaseGame {
  get name(): string {
    return "Artle";
  }

  get link(): string {
    return "https://www.nga.gov/Artle";
  }

  _buildResultRegex(): RegExp {
    return regex;
  }

  serializeResult(gameResult: RegExpMatchArray) {
    return serializeResult(gameResult);
  }

  deserialize(serializedResult: string) {
    return deserialize(serializedResult);
  }
}
