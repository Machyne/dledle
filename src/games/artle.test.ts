import { GameScore } from "../baseGame";
import { Artle } from "./artle";

const sampleWin = `
Artle #601
ðŸŽ¨ ðŸŸ¥ ðŸŸ© â¬œ â¬œ
`;

const sampleLoss = `
Artle #601
ðŸŽ¨ ðŸŸ¥ ðŸŸ¥ ðŸŸ¥ ðŸŸ¥
`;

describe("Artle", () => {
  const artle = new Artle();

  it("can parse with extra link text", () => {
    const match = artle.resultRegex.exec(sampleWin + "\nhttps://artle.wtf/");
    expect(match).not.toBeNull();
    const { score, serializedResult } = artle.serializeResult(match!);
    expect(score).toEqual(GameScore.Win);
    expect(artle.deserialize(serializedResult)).toEqual(sampleWin.trim());
  });

  for (const [name, input, expectedScore] of [
    ["sampleWin", sampleWin, GameScore.Win],
    ["sampleLoss", sampleLoss, GameScore.Loss],
  ] as Array<[string, string, GameScore]>) {
    it(`should serialize and deserialize ${name} correctly`, () => {
      const match = artle.resultRegex.exec(input);
      expect(match).not.toBeNull();
      const { score, serializedResult } = artle.serializeResult(match!);

      expect(score).toEqual(expectedScore);
      expect(artle.deserialize(serializedResult)).toEqual(input.trim());
    });
  }
});
