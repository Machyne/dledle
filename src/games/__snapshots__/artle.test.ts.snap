// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Artle serialized data should match snapshot 1`] = `
{
  "sampleLoss": {
    "score": "Loss",
    "serializedResult": "BvA",
  },
  "sampleWin": {
    "score": "Win",
    "serializedResult": "-uA",
  },
  "sampleWin trimmed": {
    "score": "Win",
    "serializedResult": "-uA",
  },
  "sampleWin with extra link text": {
    "score": "Win",
    "serializedResult": "-uA",
  },
}
`;
